### PR TITLE
Post hashcash forms

### PR DESCRIPTION
**Pull request by theus77**

### CODE REVIEW

File: Controller/ApiController.php [hunk 1]
```diff
@@ -8,20 +8,30 @@
...
 class ApiController
 {
     /**
      * @var ApiService
      */
     private $service;
-
     /**
-     * @param ApiService $service
+     * @var CsrfTokenManager
```
Comment (dameert) [inline Controller/ApiController.php]: /** @var should be a one line comment

File: Controller/ApiController.php [hunk 2]
```diff
@@ -88,6 +98,64 @@ private function treatFormRequest(Request $request, string $apiName)
...
 
+
+    private function validateHashcash(Request $request, string $csrfId, int $hashcashLevel, string $hashAlgo)
+    {
+        $hashcash = $request->headers->get('X-Hashcash');
+        if ($hashcash === null) {
+            throw new AccessDeniedHttpException('Unrecognized user');
+        }
+
+        $tokens = explode('|', $hashcash);
```
Comment (dameert) [inline Controller/ApiController.php]: Would be better to handle this in an object? To centralize the processing of the token (makes it useable for the formbundle)

File: Controller/ApiController.php [hunk 3]
```diff
@@ -88,6 +98,64 @@ private function treatFormRequest(Request $request, string $apiName)
...
+
+        if (intval($tokens[0]) < $hashcashLevel) {
```
Comment (dameert) [inline Controller/ApiController.php]: $tokens[0] would become $token->getLevel()

File: Controller/ApiController.php (hunk 1 above)
Comment (theus77) [reply Controller/ApiController.php]: Agree but the role of a controller is to validate inputs. Unless there is api controller somewhere is it's good enough for now. I'll add a TODO

Comment (theus77) [reply Controller/ApiController.php]: Refactored as recommended

File: Controller/ApiController.php (hunk 2 above)
Comment (theus77) [reply Controller/ApiController.php]: Done

File: Controller/ApiController.php (hunk 3 above)
Comment (theus77) [reply Controller/ApiController.php]: Done

File: Helper/Hashcash/Token.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+namespace EMS\ClientHelperBundle\Helper\Hashcash;
+
+
+class Token
+{
+    const DELIMITER = '|';
```
Comment (dameert) [inline Helper/Hashcash/Token.php]: I think the general rule is to put the consts of a class after the properities, check with @Davidmattei

File: Helper/Hashcash/HashcashHelper.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\Security\Csrf\CsrfTokenManager;
+
+class HashcashHelper
+{
+
+    /**
+     * @var CsrfTokenManager
```
Comment (dameert) [inline Helper/Hashcash/HashcashHelper.php]: /** @var single line :D

File: Helper/Hashcash/HashcashHelper.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        $this->csrfTokenManager = $csrfTokenManager;
+    }
+
+
+    public function validateHashcash(Request $request, string $csrfId, int $hashcashLevel = 4, string $hashAlgo = 'sha256')
+    {
+        $hashcash = $request->headers->get('X-Hashcash');
+        if ($hashcash === null) {
+            throw new AccessDeniedHttpException('Unrecognized user');
```
Comment (dameert) [inline Helper/Hashcash/HashcashHelper.php]: message should be "hashcash missing", there is not necesarrily a link with hashcash and users.

File: Helper/Hashcash/HashcashHelper.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+        $token = new Token($hashcash);
+
+        if (intval($token->getLevel()) < $hashcashLevel) {
```
Comment (dameert) [inline Helper/Hashcash/HashcashHelper.php]: make the intval part of the getLevel function?

File: Controller/ApiController.php [hunk 8]
```diff
@@ -62,47 +65,43 @@ public function document($apiName, $contentType, $ouuid)
         return $this->service->getDocument($apiName, $contentType, $ouuid)->getResponse();
     }
 
-    private function treatFormRequest(Request $request, string $apiName)
+    public function handleFormPostRequest(Request $request, string $apiName, string $contentType, ?string $ouuid, string $csrfId, string $validationTemplate, int $hashcashLevel, string $hashAlgo, bool $forceCreate = false)
```
Comment (dameert) [inline Controller/ApiController.php]: number of arguments should be reconsidered?  (General opinion, a good practice is maximum 4 parameters, possible to do more via Objects?)

File: Controller/ApiController.php [hunk 9]
```diff
@@ -17,11 +16,15 @@ class ApiController
     private $service;
 
     /**
-     * @param ApiService $service
+     * @var HashcashHelper
```
Comment (dameert) [inline Controller/ApiController.php]: /** @var single line

**Comment by dameert** [comment]
Also in general, do not resolve comments when the initial comment is not yours. It is easier to pass over a review and resolve a comment when it is read by the original poster (at last he needs to decide if it is resolved or not)

File: Controller/ApiController.php (hunk 1 above)
Comment (dameert) [reply Controller/ApiController.php]: > Agree but the role of a controller is to validate inputs. Unless there is api controller somewhere is it's good enough for now. I'll add a TODO

Think this is not on the /** @var comment? not clear what you mean

Comment (theus77) [reply Controller/ApiController.php]: Sorry, I got it now

File: Helper/Hashcash/HashcashHelper.php (hunk 6 above)
Comment (theus77) [reply Helper/Hashcash/HashcashHelper.php]: I dont want to say that the hashcash is missing as I do not want that hacker knows that they should investigate that aspect

File: Controller/ApiController.php (hunk 8 above)
Comment (theus77) [reply Controller/ApiController.php]: NO, those are route parameters

File: Helper/Hashcash/HashcashHelper.php (hunk 7 above)
Comment (theus77) [reply Helper/Hashcash/HashcashHelper.php]: good catch

File: Controller/ApiController.php (hunk 9 above)
Comment (theus77) [reply Controller/ApiController.php]: ok

File: Helper/Hashcash/HashcashHelper.php (hunk 5 above)
Comment (theus77) [reply Helper/Hashcash/HashcashHelper.php]: ok

File: Helper/Hashcash/Token.php (hunk 4 above)
Comment (theus77) [reply Helper/Hashcash/Token.php]: ok

File: Helper/Hashcash/HashcashHelper.php (hunk 6 above)
Comment (dameert) [reply Helper/Hashcash/HashcashHelper.php]: Ok, maybe no Message should be passed?

**Pull request merged**